src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs

```diff
@@ -618,21 +613,51 @@ public bool Paste()
                             case ReplSpanKind.Input:
                             case ReplSpanKind.Output:
                             case ReplSpanKind.StandardInput:
-                                InsertCode(block.Content);
+                                sb.Append(block.Content);
                                 break;
                         }
                     }
+                    code = sb.ToString();
                 }
                 else if (_window.InteractiveWindowClipboard.ContainsText())
                 {
-                    InsertCode(_window.InteractiveWindowClipboard.GetText());
+                    code = _window.InteractiveWindowClipboard.GetText();
                 }
                 else
                 {
                     return false;
                 }
 
-                return true;
+                using (var transaction = UndoHistory.CreateTransaction(InteractiveWindowResources.Paste))
+                {
+
+                    // Delete selected text if there's any and adjust caret position
+                    if (!TextView.Selection.IsEmpty)
+                    {
+                        if (CutOrDeleteSelection(isCut: false))
+                        {
+                            MoveCaretToClosestEditableBuffer();
+                        }
+                        else
+                        {
+                            return false;
```

Comment:
it returns false if the entire selection is in readonly buffer. Nothing will happen in this case (and nothing happened before this, only reading clipboard, so no operation to undo). I assume the transaction will be just disposed, is that correct?